server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregator.java

```diff
@@ -289,36 +293,210 @@ static class Compatible extends FiltersAggregator {
-                    }
-                    if (otherBucketKey != null && false == matched) {
-                        collectBucket(sub, doc, bucketOrd(bucket, docFilters.length));
+                    boolean matched = advanceAndCollectBuckets(doc, bucket, filteredScorers, filteredDocIdSetIterators, lastMatches, sub);
+                    if (otherBucketKey != null && matched == false) {
+                        collectBucket(sub, doc, bucketOrd(bucket, filteredDocIdSetIterators.length));
                     }
                 }
             };
         }
 
+        private boolean advanceAndCollectBuckets(
+            int doc,
+            long bucket,
+            Scorer[] filteredScorers,
+            DocIdSetIterator[] filteredDocIdSetIterators,
+            int[] lastMatches,
+            LeafBucketCollector sub
+        ) throws IOException {
+            boolean matched = false;
+            for (int i = 0; i < filteredDocIdSetIterators.length; i++) {
+                if (lastMatches[i] < doc) {
+                    lastMatches[i] = filteredDocIdSetIterators[i].advance(doc);
+                }
+                if (lastMatches[i] == doc) {
+                    TwoPhaseIterator twoPhase = filteredScorers[i].twoPhaseIterator();
+                    if (twoPhase == null || twoPhase.matches()) {
+                        collectBucket(sub, doc, bucketOrd(bucket, i));
+                        matched = true;
+                    }
+                }
+            }
+            return matched;
+        }
+
         final long bucketOrd(long owningBucketOrdinal, int filterOrd) {
             return owningBucketOrdinal * totalNumKeys + filterOrd;
         }
     }
 
+    /**
+     * The competitive iterator wraps the per-filter iterators, skipping rangers with no filter matches during iteration.
+     * This is achieved by storing the doc matches that each iterator returns in a priority queue that's
+     * then used to provide the next possible matches during advancing.
+     */
+    private static class CompetitiveIterator extends DocIdSetIterator {
+
+        final DocIdSetIterator[] filteredDocIdSetIterators;
+        final int[] lastMatches;
+        final DocId currentDocID;
+        final int maxDocId;
+
+        // Tracks next doc matches as they get returned from the per-filter iterators.
+        final PriorityQueue<Integer> nextDocs;
+
+        CompetitiveIterator(DocIdSetIterator[] filteredDocIdSetIterators, int[] lastMatches, DocId currentDocID, int maxDocId) {
+            this.filteredDocIdSetIterators = filteredDocIdSetIterators;
+            this.lastMatches = lastMatches;
+            this.currentDocID = currentDocID;
+            this.maxDocId = maxDocId;
+
+            nextDocs = new PriorityQueue<>(filteredDocIdSetIterators.length) {
+                @Override
+                public boolean offer(Integer e) {
+                    // Skip inserting duplicate values.
+                    if (contains(e)) {
```

Comment:
Is inserting duplicates an issue? Would this  CompetitiveIterator` should still work correctly if this wasn't done?

Reply:
Not needed any more.